### ISSUE TITLE
[Rewrite] File tree and Auto close brackets

### ISSUE THREAD

**Issue by alexmigf** [richrd/suplemon#141]
Like the title says, i suggest adding File tree and Auto close brackets.

**Comment by richrd** [comment]
Thanks for the suggestions!

The file tree is something I've wanted for a while. The problem is I'll need to do some refactoring to make it possible. The refactoring would also make it possible to show two files side by side at the same time. This will probably be the next major refactoring I'll do.

The auto close brackets is a lot easier to do, and it could be implemented as an add-on module. Making it really smart and context aware is a bit harder. Do you think it would be good enough if it worked the same way as this vim plugin: https://github.com/jiangmiao/auto-pairs#features ?

**Comment by alexmigf** [comment]
> The file tree is something I've wanted for a while. The problem is I'll need to do some refactoring to make it possible. The refactoring would also make it possible to show two files side by side at the same time. This will probably be the next major refactoring I'll do.

Great!

> The auto close brackets is a lot easier to do, and it could be implemented as an add-on module. Making it really smart and context aware is a bit harder. Do you think it would be good enough if it worked the same way as this vim plugin: https://github.com/jiangmiao/auto-pairs#features ?

Should be good.

**Comment by alexmigf** [comment]
Also you could highlight the match brackets and show a signal when a bracket isn't closed.

**Comment by richrd** [comment]
Great idea! It's probably harder than just automatically adding/removing brackets but I'll give it a try when I have time.

In the future I'll try to implement a similar plugin API as in Sublime Text (so some Sublime plugins could work in Suplemon). However that API doesn't allow changing any colors in the text so I'll need to design that part myself. I also began implementing the auto brackets, but I noticed that the Suplemon API isn't quite ready for it yet.

Refactoring the plugin API is something I'll need to do before this, but when I get it done the autobrackets feature shouldn't be hard to do.

Stay tuned!
